### ISSUE TITLE
fix(instance-ref): Instance ref in another namespace not working in add-on mode

### ISSUE THREAD

**Issue by davinkevin** [GoogleCloudPlatform/k8s-config-connector#232]
When trying to use a reference to a resource in another namespace (in case of SQL Instance and SQLDatabase or SQLUser), I have the following error.

```log
Status:
  Conditions:
    Last Transition Time:  2020-07-06T09:35:08Z
    Message:               Update call failed: error fetching live state: error getting ID for resource: error getting value from reference: error resolving reference field: error getting referenced resource from API server: sqlinstances.sql.cnrm.cloud.google.com "sql-instance-6" is forbidden: User "system:serviceaccount:cnrm-system:cnrm-controller-manager-br-24-feat-security-add-prov" cannot get resource "sqlinstances" in API group "sql.cnrm.cloud.google.com" in the namespace "infrastructure"
    Reason:                UpdateFailed
    Status:                False
    Type:                  Ready
```

**ConfigConnector Version**
1.12.1 in add-on mode following the official documentation https://cloud.google.com/config-connector/docs/how-to/install-upgrade-uninstall

**To Reproduce**

Create a SQL Instance in a namespace (`infrastructure` in the example) and `SQLUsers` or `SQLDatabase` in another namespace (named `br-24-feat-security-add-prov` in the example). 

*YAML snippets:*
```yaml
apiVersion: sql.cnrm.cloud.google.com/v1beta1
kind: SQLInstance
metadata:
  namespace: infrastructure
  name: sql-instance-6
spec:
  databaseVersion: POSTGRES_9_6
  region: europe-west1
  settings:
    tier: db-custom-4-15360
    activationPolicy: ALWAYS
    availabilityType: ZONAL
    diskSize: 100
    diskType: PD_SSD
    locationPreference:
      zone: europe-west1-b
    ipConfiguration:
      ipv4Enabled: true
      privateNetworkRef:
        external:  https://www.googleapis.com/compute/v1/projects/project-name/global/networks/default
---
apiVersion: sql.cnrm.cloud.google.com/v1beta1
kind: SQLDatabase
metadata:
  annotations:
    cnrm.cloud.google.com/project-id: project-name
  name: foo-db
  namespace: br-24-feat-security-add-prov
  labels:
    app: foo
spec:
  charset: UTF8
  collation: en_US.UTF8
  instanceRef:
    name: sql-instance-6
    namespace: infrastructure
---
apiVersion: sql.cnrm.cloud.google.com/v1beta1
kind: SQLUser
metadata:
  annotations:
    cnrm.cloud.google.com/project-id: project-name
  name: foo-user
  namespace: br-24-feat-security-add-prov
  labels:
    app: foo
spec:
  instanceRef:
    name: sql-instance-6
    namespace: infrastructure
  password:
    value: ARandomPassword
```

For information, I have other resources (without dependencies, like ComputeAddress) which are correctly applied.

I think the user just hasn't enough right to call thing created in another namespace...

**Comment by xiaobaitusi** [comment]
Hi @davinkevin , thanks for reporting the issue.

From the initial glance, I think that you are right that KCC controller KSA for `br-24-feat-security-add-prov` namespace don't have the permission to get resources in other namespace. We will look into fixing it and let you know when we have more information.

**Comment by xiaobaitusi** [comment]
Hi @davinkevin, brought this back to the team and had a discussion. Currently KCC is conservative to break the namespace boundary/isolation based on the assumption that one namespace owner might not want resources in another namespace to reference their resources by default. 

In order to allow referencing resources in `infrastructure` namespace from `br-24-feat-security-add-prov`, you can create the following `rolebinding` to enable it.
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: allow-resource-reference-from-br-24-feat-security-add-prov
  namespace: infrastructure
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cnrm-admin
subjects:
- kind: ServiceAccount
  name: cnrm-controller-manager-br-24-feat-security-add-prov
  namespace: cnrm-system
```
Let us know if that can solve your use case. For your infrastructure management, is it a common case to have resource reference cross namespaces in the namespace mode?

**Comment by davinkevin** [comment]
I've applied the RoleBinding and this is working 👍. 

In our side, this is pretty common to have `shared` resources inside a namespace and reference it from other namespaces. This allows us to reduce cost on CloudSQL for example by providing only one instance for every NS. Each ns will have their own Databases and Users defined. 

Thanks to that we are able to delete a whole ns to remove all resources attached to this "domain" without worrying about some account/db forgotten during the deletion. This is one of the major feature from our point of view in KCC.

**Comment by davinkevin** [comment]
I found another problem with this solution. I tried to apply like you proposed and it's working well but this requires any one to be able to publish manifests in the `infrastructure` namespace. In our case, we allow user to have right only in their NS. 

I tried to change the NS in the YAML you provide and I have the same error as before. Do you have any idea for this one? 

Thank for your help and support

**Comment by maqiuyujoyce** [comment]
Hi @davinkevin , thank you for your follow up. Could you provide more context about what you did? And why does it require any one to be able to publish manifests in the `infrastructure` namespace? The config we shared should be a one-time setup by `infrastructure` ns admin, and users shouldn't need to do anything about it.

**Comment by davinkevin** [comment]
The solution you provide has to be published in the `infrastructure` namespace. I tried to publish it in another ns (for example in `br-24-feat-security-add-prov`) and this leads to the same error as before. So I conclude this has to be done in the NS where the instance is declared. 

We use our namespace to segregate every envs. In our case, we are deploying a new set of application in a new namespace for every branch in our source code... which requires to publish a  `RoleBinding` for every branch (and `namespace`).

We leverage KCC to simplify our deployment strategy from dev to production, and be able to test every iteration, even during development, with the same configuration used by the production. 

I've seen the changelog of the version `1.6.0` the following element:

> Operator improvement: add support for cluster-mode set-ups, which allows users to use one Google Service Account for all namespaces in their cluster. This is very similar to the traditional "Workload Identity" installation set-up.

Is it something which could be helpful in my use-case ? Especially in the AddOn mode? In the previous version (manually installed with WorkLoad Identity), we didn't have to play with this rolebinding nor `gcloud iam add-…` commands, which was good for us 👍. 

If you want to talk about it through a meet visio, let me know 😉

**Comment by jcanseco** [comment]
> I found another problem with this solution. I tried to apply like you proposed and it's working well but this requires any one to be able to publish manifests in the infrastructure namespace. In our case, we allow user to have right only in their NS.

Hi @davinkevin, gotcha, I think I understand the problem: basically the solution we provided [1] requires users from every namespace who needs to be able to reference resources in the `infrastructure` namespace to apply the provided `RoleBinding` in the `infrastructure` namespace, which they would not have the permission to do so. Can you confirm if my understanding is correct?

[1] The solution we provided (repurposed to be re-usable for other namespaces that need to be able to reference resources in the `infrastructure` namespace):
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: allow-resource-reference-from-[NAMESPACE]
  namespace: infrastructure
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cnrm-admin
subjects:
- kind: ServiceAccount
  name: cnrm-controller-manager-[NAMESPACE]
  namespace: cnrm-system
```

> I've seen the changelog of the version 1.6.0 the following element:

>> Operator improvement: add support for cluster-mode set-ups, which allows users to use one Google Service Account for all namespaces in their cluster. This is very similar to the traditional "Workload Identity" installation set-up.

> Is it something which could be helpful in my use-case ? Especially in the AddOn mode? In the previous version (manually installed with WorkLoad Identity), we didn't have to play with this rolebinding nor gcloud iam add-… commands, which was good for us +1.

Yes, potentially. If you don't require namespaced-mode (i.e. if you do not require the ability to use different Google Service Accounts per namespace), and you are ok with using just one Google Service Account for all your namespaces, then yes, cluster-mode would fit your use-case -- and would eliminate the need to create `RoleBinding`'s to allow for cross-namespace references. Like you said, if you were already using a Workload Identity installation before and that worked for you, then the operator in cluster-mode is most likely going to work for you as well.

Although please note that this latest version of the KCC operator (v1.15.0), which adds support for cluster-mode has not yet been picked up by the GKE add-on. You _can_ install the operator as a standalone, but if you wish to stick with using the add-on, then you'll have to wait another 2 weeks.

Do you think using the operator in cluster-mode will work for you? We plan to update our docs to go over how to use cluster-mode with the operator in the coming weeks, but we can give you a quick overview here if you believe it fits your use-case.

**Comment by davinkevin** [comment]
You understand well the problem 👍.

And I'm glad to here the 1.15 will provide a solution simpler for our use case. I can wait for the official release. Currently we are working with the manually installed operator... I've discovered the limitation and problems during a migration to the add on mode... and at the end, we choose to rollback to the old solution, waiting for a solution.. exactly what you provide.

I will wait the release on GKE and provide you a feedback. I let you choose if this issue should be close or not.

**Comment by jcanseco** [comment]
Gotcha, sounds great @davinkevin, we greatly appreciate the feedback and detailed explanations you provide (this thread being an example). I can leave this issue open for now in order to confirm later that the fix works for you.

**Comment by toumorokoshi** [comment]
Hi! I'm closing this issue as we have a fix, and 1.15 should be available in GKE addOns at this point. Please feel free to re-open if you run into furhher issues. Thanks!
